components/operations-controller/internal/webhook/client.go

```diff
@@ -0,0 +1,207 @@
+ * Copyright 2020 The Compass Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package webhook
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/kyma-incubator/compass/components/director/pkg/correlation"
+	"github.com/kyma-incubator/compass/components/operations-controller/internal/auth"
+	"github.com/kyma-incubator/compass/components/operations-controller/internal/log"
+
+	"github.com/kyma-incubator/compass/components/director/pkg/graphql"
+	web_hook "github.com/kyma-incubator/compass/components/director/pkg/webhook"
+	"github.com/pkg/errors"
+)
+
+type client struct {
+	httpClient *http.Client
+}
+
+func NewClient(httpClient *http.Client) *client {
+	return &client{
+		httpClient: httpClient,
+	}
+}
+
+func (c *client) Do(ctx context.Context, request *Request) (*web_hook.Response, error) {
+	var err error
+	webhook := request.Webhook
+
+	var method string
+	url := webhook.URL
+	if webhook.URLTemplate != nil {
+		resultURL, err := request.Object.ParseURLTemplate(webhook.URLTemplate)
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to parse webhook URL")
+		}
+		url = resultURL.Path
+		method = *resultURL.Method
+	}
+
+	if url == nil {
+		return nil, errors.New("missing webhook url")
+	}
+
+	body := []byte("{}")
+	if webhook.InputTemplate != nil {
+		body, err = request.Object.ParseInputTemplate(webhook.InputTemplate)
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to parse webhook input body")
+		}
+	}
+
+	headers := http.Header{}
+	if webhook.HeaderTemplate != nil {
+		headers, err = request.Object.ParseHeadersTemplate(webhook.HeaderTemplate)
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to parse webhook headers")
+		}
+	}
+
+	ctx = correlation.SaveCorrelationIDHeaderToContext(ctx, webhook.CorrelationIDKey, &request.CorrelationID)
+
+	req, err := http.NewRequestWithContext(ctx, method, *url, bytes.NewBuffer(body))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header = headers
+
+	if webhook.Auth != nil {
+		ctx = auth.SaveToContext(ctx, webhook.Auth.Credential)
+		req = req.WithContext(ctx)
+	}
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	responseObject, err := parseResponseObject(resp)
+	if err != nil {
+		return nil, err
+	}
+
+	log.C(ctx).Info(fmt.Sprintf("Webhook response object: %v", *responseObject))
+
+	response, err := responseObject.ParseOutputTemplate(webhook.OutputTemplate)
```

Comment:
We cannot enforce clients to always register these templates as that would be backward incompatible, so theoretically, you're right, it is possible for it it be nil here. Since we cannot have validation upon webhook registration, we'll add additional checks here.